src/Shared/UnitTests/TestEnvironment.cs

```diff
@@ -0,0 +1,358 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Shared;
+using Microsoft.Build.UnitTests;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Build.Engine.UnitTests
+{
+    public class TestEnvironment : IDisposable
+    {
+        /// <summary>
+        ///     List of test invariants to assert value does not change.
+        /// </summary>
+        private readonly List<TestInvariant> _invariants = new List<TestInvariant>();
+
+        /// <summary>
+        ///     List of test variants which need to be reverted when the test completes.
+        /// </summary>
+        private readonly List<TransientTestState> _variants = new List<TransientTestState>();
+
+        private readonly ITestOutputHelper _output;
+
+        public static TestEnvironment Create(ITestOutputHelper output = null, bool ignoreBuildErrorFiles = false)
+        {
+            var env = new TestEnvironment(output ?? new DefaultOutput());
+
+            // In most cases, if MSBuild wrote an MSBuild_*.txt to the temp path something went wrong.
+            if (!ignoreBuildErrorFiles)
+                env.WithInvariant(new BuildFailureLogInvariant());
+
+            return env;
+        }
+
+        private TestEnvironment(ITestOutputHelper output)
+        {
+            _output = output;
+            SetDefaultInvariant();
+        }
+
+        /// <summary>
+        ///     Revert / cleanup variants and then assert invariants.
+        /// </summary>
+        public void Dispose()
+        {
+            // Reset test variants
+            foreach (var variant in _variants)
+                variant.Revert();
+
+            // Assert invariants
+            foreach (var item in _invariants)
+                item.AssertInvariant(_output);
+        }
+
+        /// <summary>
+        ///     Evaluate the test with the given invariant.
+        /// </summary>
+        /// <param name="invariant">Test invariant to assert unchanged on completion.</param>
+        public T WithInvariant<T>(T invariant) where T : TestInvariant
+        {
+            _invariants.Add(invariant);
+            return invariant;
+        }
+
+        /// <summary>
+        ///     Evaluate the test with the given transient test state.
+        /// </summary>
+        /// <returns>Test state to revert on completion.</returns>
+        public T WithTransientTestState<T>(T transientState) where T : TransientTestState
+        {
+            _variants.Add(transientState);
+            return transientState;
+        }
+
+        #region Common test variants
+
+        private void SetDefaultInvariant()
+        {
```

Comment:
Not a bad idea in combination with setting a unique temp folder per test. Without that though too much possibility of external things affecting that.